### PR TITLE
Enable per-token classification in RoBERTa

### PR DESCRIPTION
**Pull request by prihoda**
# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
- [x] Did you read the [contributor guideline](https://github.com/pytorch/fairseq/blob/master/CONTRIBUTING.md)?
- [ ] Did you make sure to update the docs?   
- [x] Did you write any new necessary tests?  

## What does this PR do?
Fixes https://github.com/pytorch/fairseq/issues/1708

### CODE REVIEW

File: fairseq/models/roberta/model.py [hunk 1]
```diff
@@ -230,15 +231,18 @@ def forward(self, features, masked_tokens=None, **kwargs):
...
+        self.sequence_tagging = sequence_tagging
 
     def forward(self, features, **kwargs):
-        x = features[:, 0, :]  # take <s> token (equiv. to [CLS])
+        x = features
+        if not self.sequence_tagging:
+            x = x[:, 0, :]  # take <s> token (equiv. to [CLS])
         x = self.dropout(x)
         x = self.dense(x)
         x = self.activation_fn(x)
```
Comment (prihoda) [inline fairseq/models/roberta/model.py]: @myleott Would it make sense to add `layer_norm` here just like in the `RobertaLMHead`? https://github.com/pytorch/fairseq/pull/1709/files#diff-ad4f875b91a61dcc1511c93c8e38cc4bL224

File: fairseq/tasks/sequence_tagging.py [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+            )
+            dataset['net_input'].update(
+                prev_output_tokens=prev_tokens_dataset,
+            )
+
+        if not self.args.regression_target:
+            label_dataset = make_dataset('label', self.label_dictionary)
+            if label_dataset is not None:
+                dataset.update(
+                    target=RightPadDataset( # append -1 as padding token
```
Comment (prihoda) [inline fairseq/tasks/sequence_tagging.py]: @myleott I'm removing the `eos` token and padding the dataset with `-1` to keep the class dictionary in the original range (`0,1,2,...`) and to enable masking. Is there a better way?

File: fairseq/tasks/sequence_tagging.py [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+                                ),
+                                offset=-self.label_dictionary.nspecial,
+                            ),
+                            token=-1,
+                        ),
+                        pad_idx=-1
+                    )
+                )
+        else:
+            raise NotImplementedError()
```
Comment (prihoda) [inline fairseq/tasks/sequence_tagging.py]: @myleott Can I keep this here or shall I rather remove the regression option before it's implemented?

File: fairseq/criterions/sequence_tagging.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+            **sample['net_input'],
+            features_only=True,
+            classification_head_name=self.args.classification_head_name,
+        )
+        targets = model.get_targets(sample, [logits]).view(-1)
+        sample_size = sample['ntokens'] - sample['target'].size(0) # number of tokens without eos
+
+        if not self.args.regression_target:
+            logits = logits.view(-1, logits.size(-1))
+            loss = F.nll_loss(
```
Comment (prihoda) [inline fairseq/criterions/sequence_tagging.py]: @myleott Can you please validate this loss calculation?

**Comment by prihoda** [comment]
@myleott What else is needed to make this mergeable? Do I need to create a dedicated RoBERTa example README for sequence tagging?

File: fairseq/criterions/sequence_tagging.py [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+
+
+@register_criterion('sequence_tagging')
+class SequenceTaggingCriterion(FairseqCriterion):
+
+    @staticmethod
+    def add_args(parser):
+        # fmt: off
+        parser.add_argument('--classification-head-name',
+                            default='sentence_classification_head',
```
Comment (myleott) [inline fairseq/criterions/sequence_tagging.py]: `sequence_tagging_head`

File: fairseq/criterions/sequence_tagging.py (hunk 4 above)
Comment (myleott) [reply fairseq/criterions/sequence_tagging.py]: Looks good, but it'd be great if you could test this on a real sequence tagging task first.

File: fairseq/tasks/sequence_tagging.py [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+        input1 = make_dataset('input1', self.source_dictionary)
+
+        if self.args.init_token is not None:
+            input0 = PrependTokenDataset(input0, self.args.init_token)
+
+        if input1 is None:
+            src_tokens = input0
+        else:
+            if self.args.separator_token is not None:
+                input1 = PrependTokenDataset(input1, self.args.separator_token)
```
Comment (myleott) [inline fairseq/tasks/sequence_tagging.py]: Do we ever have two inputs for this kind of task?

File: fairseq/tasks/sequence_tagging.py [hunk 7]
```diff
@@ -0,0 +1,242 @@
...
+        parser.add_argument('--init-token', type=int, default=None,
+                            help='add token at the beginning of each batch item')
+        parser.add_argument('--separator-token', type=int, default=None,
+                            help='add separator token between inputs')
+        parser.add_argument('--regression-target', action='store_true', default=False)
+        parser.add_argument('--no-shuffle', action='store_true', default=False)
+        parser.add_argument('--truncate-sequence', action='store_true', default=False,
+                            help='truncate sequence to max-positions')
+        parser.add_argument('--add-prev-output-tokens', action='store_true', default=False,
+                            help='add prev_output_tokens to sample, used for encoder-decoder arch')
```
Comment (myleott) [inline fairseq/tasks/sequence_tagging.py]: We can probably remove `--separator-token` and `--add-prev-output-tokens`

File: fairseq/tasks/sequence_tagging.py [hunk 8]
```diff
@@ -0,0 +1,242 @@
...
+        if self.args.add_prev_output_tokens:
+            prev_tokens_dataset = RightPadDataset(
+                RollDataset(src_tokens, 1),
+                pad_idx=self.dictionary.pad(),
+            )
+            dataset['net_input'].update(
+                prev_output_tokens=prev_tokens_dataset,
+            )
+
+        if not self.args.regression_target:
```
Comment (myleott) [inline fairseq/tasks/sequence_tagging.py]: Since we don't support add_prev_output_tokens or regression_target, shall we fold this into the `dataset` dictionary definition above?

File: fairseq/tasks/sequence_tagging.py (hunk 2 above)
Comment (myleott) [reply fairseq/tasks/sequence_tagging.py]: This seems fine, but please add a comment in the code

File: fairseq/tasks/sequence_tagging.py [hunk 9]
```diff
@@ -0,0 +1,242 @@
...
+            label_dataset = make_dataset('label', self.label_dictionary)
```
Comment (myleott) [inline fairseq/tasks/sequence_tagging.py]: We should add some kind of check that the lengths of label dataset match the length of the src_tokens dataset.

How about something like this:
```
dataset = {
    'id': IdDataset(),
    ...
    'target': label_dataset,
    '_assert_lengths_match': AssertSameLengthDataset(src_tokens, label_dataset),
}
```

AssertSameLengthDataset can just return `1` if the lengths of each item match, otherwise raise an exception.

File: fairseq/models/roberta/model.py [hunk 10]
```diff
@@ -112,7 +112,7 @@ def forward(self, src_tokens, features_only=False, return_all_hiddens=False, cla
             x = self.classification_heads[classification_head_name](x)
         return x, extra
 
-    def register_classification_head(self, name, num_classes=None, inner_dim=None, **kwargs):
+    def register_classification_head(self, name, num_classes=None, sequence_tagging=False, inner_dim=None, **kwargs):
```
[the remark this answers is not in the document]
Comment (prihoda) [reply fairseq/models/roberta/model.py]: And these would be stored in `self.classification_heads` or in separate `self.sequence_tagging_heads`?

File: fairseq/criterions/sequence_tagging.py [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+
+        Returns a tuple with three elements:
+        1) the loss
+        2) the sample size, which is used as the denominator for the gradient
+        3) logging outputs to display while training
+        """
+        assert (
+            hasattr(model, 'classification_heads')
+            and self.args.classification_head_name in model.classification_heads
+        ), 'model must provide sentence classification head for --criterion=sequence_tagging'
```
Comment (prihoda) [inline fairseq/criterions/sequence_tagging.py]: @myleott How shall I handle this assertion when we are performing sequence tagging? Shall I check that either is present?

File: fairseq/criterions/sequence_tagging.py (hunk 4 above)
Comment (prihoda) [reply fairseq/criterions/sequence_tagging.py]: I added an example with a README 😉 https://github.com/pytorch/fairseq/pull/1709/files#diff-dcc107dc4ee613a610d8174151f0afc4

This made me realize that it might be tricky to use sequence tagging with BPE, since you will have to provide a label for each input BPE token. I clarified that in the README, hope that is enough.

I also added `predict_class` and `predict_tags` convenience methods and used them in the classification and tagging READMEs.

Comment (prihoda) [reply fairseq/criterions/sequence_tagging.py]: By the way, shouldn't I name this task `sentence_tagging` instead of `sequence_tagging`, since we already have `sentence_prediction` and `sentence_ranking`?

File: fairseq/tasks/sequence_tagging.py (hunk 6 above)
Comment (prihoda) [reply fairseq/tasks/sequence_tagging.py]: Good point, I removed this functionality

**Comment by prihoda** [comment]
@myleott I added an example with a README 😉 https://github.com/pytorch/fairseq/pull/1709/files#diff-dcc107dc4ee613a610d8174151f0afc4

This made me realize that it might be tricky to use sequence tagging with BPE, since you will have to provide a label for each input BPE token. I clarified that in the README, hope that is enough.

I also added `predict_class` and `predict_tags` convenience methods and used them in the classification and tagging READMEs.

By the way, shouldn't I name this task `sentence_tagging` instead of `sequence_tagging`, since we already have `sentence_prediction` and `sentence_ranking`?

**Comment by prihoda** [comment]
@myleott I updated the code to reflect #1729

**Comment by prihoda** [comment]
@myleott I rebased the PR on develop. You mentioned that you want to add a separate `register_sequence_tagging_head` function, to make it clearer what models support sequence tagging vs. sentence-level classification. Should these be stored in `self.classification_heads` or in separate `self.sequence_tagging_heads`?

**Comment by avi-jit** [comment]
@prihoda I'm trying out your branch feature/classify_per_token on a custom sequence tagging task, but am running into an error: `has_marked_unused_parameters_ INTERNAL ASSERT FAILED at /pytorch/torch/csrc/distributed/c10d/reducer.cpp:327` which resembles the issues pytorch/pytorch#32490 and pytorch/pytorch#31035.

Any help would be appreciated in debugging. I didn't want to paste the lengthy log here unless it's relevant to this issue. Thanks in advance :)

**Comment by prihoda** [comment]
Hi @avi-jit, let me know if you find out how this could be related to this branch, at this moment I'm not sure how to find the cause of your issue. Can you first try training using a simpler `sentence_prediction` task?

**Comment by avi-jit** [comment]
sure, i'll try that @prihoda. Could you confirm if feature/classify_per_token is the correct branch to be working with?

**Comment by prihoda** [comment]
@avi-jit yes that's the one

**Comment by apyspo** [comment]
Could you please re-open this one?

File: fairseq/models/roberta/model.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (prihoda) [reply fairseq/models/roberta/model.py:115]: @myleott Any thoughts? Thanks

**Comment by apyspo** [comment]
Could you please also share preprocess_tagging.sh mentioned in the README file

**Comment by prihoda** [comment]
Hi @apyspo I tried to look for this file but can't seem to find it, you would probably need to write the script yourself. It transforms the downloaded files into separate input and label files, one sentence per line.

**Comment by apyspo** [comment]
> 
Thank you @prihoda , I have written a script and transformed into one sentence per line similar to top ten sample lines in the README file.

**Comment by prihoda** [comment]
Hi @apyspo will you try to finish this pull request? We could ask the main contributors what's needed to have this merged.
